### PR TITLE
Use constraints from $JUJU_TEST_CONSTRAINTS if specified.

### PR DESCRIPTION
**Pull request by dannf**

### CODE REVIEW

**Comment by marcoceppi** [comment]
Could you please rebase this against master? There are some merge conflicts.

**Comment by tvansteenburgh** [comment]
Isn't this identical to https://github.com/juju/amulet/pull/83 ?

**Comment by marcoceppi** [comment]
Seems like it, but #83 was closed by @dannf still has merge conflicts though

**Comment by tvansteenburgh** [comment]
Ah, my bad, I thought #83 was merged.

**Comment by dannf** [comment]
I've rebased against master. And yeah - sorry for the withdrawal/resubmit - i noticed some excess whitespace the first time.

File: amulet/deployer.py [hunk 1]
```diff
@@ -136,6 +136,15 @@ def add(self, service_name,
         if placement is not None:
             service['to'] = placement
 
+        if constraints is None and 'JUJU_TEST_CONSTRAINTS' in os.environ:
```
Comment (marcoceppi) [inline amulet/deployer.py]: This seems fine, but is there a case where JUJU_TEST_CONSTRAINTS should be merged with the constraints provided by the author? Either to override or append?

Comment (dannf) [inline amulet/deployer.py]: My thought is that the test author's constraints should take precedent, but it does make sense that environmental constraints could be appended. My reasoning is that the test author should only be specifying constraints that are required for the deployment, or the specific test being executed - overriding those would seemingly only lead to false negatives.

If you approve of that logic, I can update my branch. I also wonder if we shouldn't change the environment variable name to make the interface's behavior more obvious - e.g. JUJU_TEST_EXTRA_CONSTRAINTS? (or maybe AMULET_EXTRA_CONSTRAINTS? not sure what the best namespace is).

Comment (marcoceppi) [inline amulet/deployer.py]: Using `JUJU_TEST` prefix is best, as amulet is just meant to make test writing easier, but it's not the only way to write tests. I'd be okay with extra constraints, I could see cases where the testing infrastructure may want to set variables for deployment.

I guess if we knew the use case for this it'd be better to decide the best name/behavior for it in Amulet.

Comment (tvansteenburgh) [inline amulet/deployer.py]: My $.02:

1. JUJU_TEST_CONSTRAINTS is a good name.
2. The constraints from the env var should always be applied if they exist. The test author's constraints should be applied over the top of them, so they take precedence in the case of a conflict.

Comment (dannf) [inline amulet/deployer.py]: @marcoceppi my use case is just the one mentioned in the commit log - that is, we have MAAS cluster with multiple architectures, but we'd like to be able to limit the test to running across a single architecture to e.g. verify arm64 compatibility.

Comment (marcoceppi) [inline amulet/deployer.py]: @dannf sure, so if a charm author sets an architecture your settings won't take place. There should probably be two options here

```
JUJU_TEST_CONSTRAINTS
JUJU_TEST_EXTRA_CONSTRAINTS
````

The first one is likely what you want, it'll overwrite and takes priority for constraints regardless if the author has created supplied one or not. The second uses the logic you've defined now.

So replacing `JUJU_TEST_CONSTRAINTS` with `JUJU_TEST_EXTRA_CONSTRAINTS`, then adding another if block for if os.environ has `JUJU_TEST_CONSTRAINTS` and those overriding/setting constraints for the service.

@tvansteenburgh opinions?

Comment (tvansteenburgh) [inline amulet/deployer.py]: @marcoceppi I'd rather keep it simple and just have JUJU_TEST_CONSTRAINTS, with author's constraints always overriding them.

Comment (marcoceppi) [inline amulet/deployer.py]: @tvansteenburgh That doesn't satisfy my requirement of "let me test on different architectures" if the author explicitly puts an arch constraint

Comment (marcoceppi) [inline amulet/deployer.py]: Though, I don't actually have that use case at the moment - just inferring it from @dannf

Comment (tvansteenburgh) [inline amulet/deployer.py]: If the author includes an arch constraint, it's likely b/c they know the charm only works on that arch anyway.

Comment (marcoceppi) [inline amulet/deployer.py]: Yes, I get that, but there's no way to verify that. Trust but verify. End of the day, if this satisfies @dannf's needs I'm good with it. If we ever come up to a point where we need to override authors constraints a `JUJU_TEST_CONSTRAINTS_OVERRIDES` flag, or similar, could be added

Comment (tvansteenburgh) [inline amulet/deployer.py]: +1, I understand your point, I just don't want to add code that solves a problem that nobody has. YAGNI and all that. Plus, arguing on github is fun.

**Comment by dannf** [comment]
fyi, I've implemented the mixing of the environmental and test-explicit constraints in #85.

**Pull request merged**